File: Tools/setuptool/HemeLbSetupTool/Model/Neighbours.py
"""HemeLB's ordering of the vectors linking a site to its 26 neighbours. 
"""
import numpy as np

neighbours = np.array(((-1,-1,-1),
                       (-1,-1, 0),
                       (-1,-1,+1),
                       (-1, 0,-1),
                       (-1, 0, 0),
                       (-1, 0,+1),
                       (-1,+1,-1),
                       (-1,+1, 0),
                       (-1,+1,+1),
                       ( 0,-1,-1),
                       ( 0,-1, 0),
                       ( 0,-1,+1),
                       ( 0, 0,-1),
                       # ( 0, 0, 0),
                       ( 0, 0,+1),
                       ( 0,+1,-1),
                       ( 0,+1, 0),
                       ( 0,+1,+1),
                       (+1,-1,-1),
                       (+1,-1, 0),
                       (+1,-1,+1),
                       (+1, 0,-1),
                       (+1, 0, 0),
                       (+1, 0,+1),
                       (+1,+1,-1),
                       (+1,+1, 0),
                       (+1,+1,+1)), dtype=int)
inverses = np.array((25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0))
assert np.all(neighbours + neighbours[inverses] == 0)

norm2 = np.sum(neighbours**2, axis=1)
norm = np.sqrt(norm2)

File: Tools/setuptool/HemeLbSetupTool/Model/OldGeneration/Neighbours.cpp
//
// Copyright (C) University College London, 2007-2012, all rights reserved.
//
// This file is part of HemeLB and is CONFIDENTIAL. You may not work
// with, install, use, duplicate, modify, redistribute or share this
// file, or any part thereof, other than as allowed by any agreement
// specifically made by you with University College London.
//

#include "Neighbours.h"

#include "io/formats/geometry.h"
// shortcut to geometry class
using hemelb::io::formats::geometry;

void Neighbours::Init() {
	// copy in the vectors of the neighbourhood
	Neighbours::vectors = geometry::GetNeighbourhood();

	//unsigned int iLater = 0;
	for (unsigned int i = 0, iLater = 0; i < Neighbours::n; ++i) {
		Neighbours::norms[i] = std::sqrt(Neighbours::vectors[i].GetMagnitudeSquared());

		// Figure out if this takes us to a later site
		if (Neighbours::vectors[i][0] > 0) {
			// yes
			Neighbours::laterNeighbourIndices[iLater] = i;
			iLater++;
		} else if (Neighbours::vectors[i][0] == 0) {
			// Maybe
			if (Neighbours::vectors[i][1] > 0) {
				// yes
				Neighbours::laterNeighbourIndices[iLater] = i;
				iLater++;
			} else if (Neighbours::vectors[i][1] == 0) {
				// Maybe
				if (Neighbours::vectors[i][2] > 0) {
					// yes
					Neighbours::laterNeighbourIndices[iLater] = i;
					iLater++;
				} else {
					// No
				}
			} else {
				// No
			}
		} else {
			// No
		}

		// Complete the table of inverse indices.
		for (unsigned int j = i; j < Neighbours::n; ++j) {
			if (Neighbours::vectors[i] == -Neighbours::vectors[j]) {
				Neighbours::inverses[i] = j;
				Neighbours::inverses[j] = i;
				break;
			}
		}
	}
}

std::vector<unsigned int> Neighbours::laterNeighbourIndices(Neighbours::nLater);
std::vector<Index> Neighbours::vectors(Neighbours::n);
std::vector<unsigned int> Neighbours::inverses(Neighbours::n);
std::vector<double> Neighbours::norms(Neighbours::n);

